service/interpreter/stateExecutionCounter.go

```diff
@@ -135,27 +137,61 @@ func (e *StateExecutionCounter) MarkStateExecutionCompleted(state iwfidl.StateMo
 		case iwfidl.DISABLED:
 			return nil
 		case iwfidl.ENABLED_FOR_ALL:
-			e.decreaseStateIdCurrentlyExecutingCounts(state)
+			e.decreaseStateIdCurrentlyExecutingCounts(currentState)
+			shouldSkipUpsert := determineIfShouldSkipUpsert(currentState, nextStates)
+			if shouldSkipUpsert {
+				return nil
+			}
 		case iwfidl.ENABLED_FOR_STATES_WITH_WAIT_UNTIL:
 			fallthrough
 		default:
 			if compatibility.GetSkipWaitUntilApi(&options) {
 				return nil
 			} else {
-				e.decreaseStateIdCurrentlyExecutingCounts(state)
+				e.decreaseStateIdCurrentlyExecutingCounts(currentState)
+				shouldSkipUpsert := determineIfShouldSkipUpsert(currentState, nextStates)
+				if shouldSkipUpsert {
+					return nil
+				}
 			}
 		}
 	} else {
 		if config.GetDisableSystemSearchAttribute() {
 			return nil
 		} else {
-			e.decreaseStateIdCurrentlyExecutingCounts(state)
+			e.decreaseStateIdCurrentlyExecutingCounts(currentState)
 		}
 	}
 
 	return e.updateStateIdSearchAttribute()
 }
 
+func determineIfShouldSkipUpsert(currentState iwfidl.StateMovement, nextStates []iwfidl.StateMovement) bool {
```

Comment:
I thought the signature would have pass `ExecutingStateIdMode`.
I think the currentState can be omitted. 

```
func determineIfShouldSkipUpsert(mode ExecutingStateIdMode, nextStates ) bool{
  // based on the mode
  // For ENABLED_FOR_ALL, return true as long as any of next states is going to a next state(not closing state)
  // For ENABLED_FOR_STATES_WITH_WAIT_UNTIL, in addition to above, also check the state has waitUntil.
}
```
StateMovement may not go to any state, because we are doing a hack in iwf (I really regret that) 😨 
We use some special StateIDs to prepresent gracefulComplete, deadEnd etc. See https://github.com/indeedeng/iwf/blob/main/service/const.go#L68
And how it's being used in https://github.com/indeedeng/iwf/blob/main/service/interpreter/workflowImpl.go#L373

Reply:
> I thought the signature would have pass ExecutingStateIdMode.
I'm not passing it as I only invoke determineIfShouldSkipUpsert when needed based on `ExecutingStateIdMode`

currentState is needed if we want to determine when a state is looping back, I believe

Reply:
> currentState is needed if we want to determine when a state is looping back, I believe
for that case, I belive what you have in  `updateStateIdSearchAttribute` baesd on comparing with GetSearchAttributes will cover it?

Reply:
Basically, if `MarkStateExecutionCompleted` not delete the SA, next time when `MarkStateExecutingIfNotYet`->`updateStateIdSearchAttribute ` will then skip it because the SA is still there